File: Tang/Tang/Content/General/Categories/UIKitCategory/UIViewController+Storyboard.m
//
//  UIViewController+Storyboard.m
//
//
//  Created by Normal on 15/11/17.
//  Copyright © 2015年 Bo Wang. All rights reserved.
//

#import "UIViewController+Storyboard.h"

// 使用 YYCache 或 PINCache 时将把storyboard名进行本地缓存
#if (__has_include(<YYCache/YYCache.h>) || __has_include("YYCache.h"))
    #import <YYCache/YYCache.h>
    typedef YYCache _YGCache;
#elif (__has_include(<PINCache/PINCache.h>) || __has_include("PINCache.h"))
    #import <PINCache/PINCache.h>
    typedef PINCache _YGCache;
#else
    typedef NSCache _YGCache;

    @interface _YGCache (_category)
    @end
    @implementation _YGCache (_category)
    - (instancetype)initWithName:(NSString *)name
    {
        _YGCache *cache = [self init];
        cache.name = name;
        return cache;
    }
    @end
#endif

@implementation UIViewController (Storyboard)

+ (NSDictionary<NSString *,NSSet<NSString *> *> *)storyboardIdentifierMap
{
    static NSDictionary *kViewControllerIdentifierToClassMap;
    static dispatch_once_t onceToken;
    dispatch_once(&onceToken, ^{
        NSMutableDictionary *tmp = [NSMutableDictionary dictionary];
        NSString *resourcePath = [NSBundle mainBundle].resourcePath;
        NSArray *list = [NSBundle pathsForResourcesOfType:@"storyboardc" inDirectory:resourcePath];
        for (NSString *path in list) {
            NSString *name = [[path lastPathComponent] stringByDeletingPathExtension];
            NSString *realName = [[name componentsSeparatedByString:@"~"] firstObject];
            NSBundle *bundle = [NSBundle bundleWithPath:path];
            NSDictionary *nibNames = [bundle objectForInfoDictionaryKey:@"UIViewControllerIdentifiersToNibNames"];
            tmp[realName] = [NSSet setWithArray:nibNames.allKeys];
        }
        kViewControllerIdentifierToClassMap = tmp;
    });
    return kViewControllerIdentifierToClassMap;
}

FOUNDATION_EXTERN NSString *const kAppBundleID;

+ (_YGCache *)cache
{
    static _YGCache *cache;
    static dispatch_once_t onceToken;
    dispatch_once(&onceToken, ^{
        NSString *bundleid = [[NSBundle mainBundle] objectForInfoDictionaryKey:(__bridge NSString *)kCFBundleIdentifierKey];
        NSString *cacheName = [bundleid stringByAppendingString:@".UIViewController+Storyboard"];
        cache = [[_YGCache alloc] initWithName:cacheName];
    });
    return cache;
}

/**
 *  尝试从以storyboardName为名的storyboard中取
 *
 *  @param storyboardName storyboard文件名
 *  @param identifier     查找的对象id
 *
 *  @return 查找到的UIViewController实例。如果没有找到，返回nil
 */
+ (instancetype)tryTakeOutInstanceFromStoryboardNamed:(NSString *)storyboardName identifier:(NSString *)identifier
{
    if (!storyboardName || !identifier) return nil;
    
    /**
     *  这里需要捕获异常，否则程序会crash
     */
    @try {
        static SEL sel;
        static dispatch_once_t onceToken;
        dispatch_once(&onceToken, ^{
            //identifierToNibNameMap 是UIStoryboard的一个私有属性
            NSString *selStr = [NSString stringWithFormat:@"%@%@%@%@%@",@"identifier",@"To",@"Nib",@"Name",@"Map"];
            sel = NSSelectorFromString(selStr);
        });

        UIStoryboard *sb = [UIStoryboard storyboardWithName:storyboardName bundle:[NSBundle mainBundle]];
        NO_WARNING_BEGIN(-Warc-performSelector-leaks)
        id obj = [sb performSelector:sel];
        NO_WARNING_END
        if (obj && [obj isKindOfClass:[NSDictionary class]] && [[(NSDictionary *)obj allValues] containsObject:identifier]) {
            UIViewController *vc = [sb instantiateViewControllerWithIdentifier:identifier];
            return vc;
        }
        return nil;
    }
    @catch (NSException *exception) {
        NSLog(@"%@",exception);
        return nil;
    }
}

+ (instancetype)instanceFromStoryboardWithIdentifier:(NSString *)identifier
{
    if (!identifier) return nil;
    
    // 取缓存的storyboard名
    _YGCache *cache = [self cache];
    NSString *cacheStoryboardName = (NSString *)[cache objectForKey:identifier];
    if (cacheStoryboardName) {
        id vc = [self tryTakeOutInstanceFromStoryboardNamed:cacheStoryboardName identifier:identifier];
        if (vc) {
            return vc;
        }
        // 缓存的name不再有效
        [cache removeObjectForKey:identifier];
    }
    // 未缓存，遍历storyboard文件名列表，开始尝试取出实例。
    NSDictionary *map = [self storyboardIdentifierMap];
    for (NSString *name in map) {
        NSSet *identifierList = map[name];
        NSString *theIdentifier = [[identifier componentsSeparatedByString:@"."] lastObject];
        if ([identifierList containsObject:theIdentifier]) {
            UIViewController *instance = [self tryTakeOutInstanceFromStoryboardNamed:name identifier:theIdentifier];
            if (instance) {
                // 成功获取实例后，对storyboard名进行缓存
                [cache setObject:name forKey:identifier];
                return instance;
            }
        }
    }
    return [[self alloc] init];
}

+ (instancetype)instanceFromStoryboard
{
    return [self instanceFromStoryboardWithIdentifier:NSStringFromClass([self class])];
}

@end

File: Tang/Tang/General/Extension/UIKit/UIViewController+Storyboard.swift
//
//  UIViewController+Storyboard.swift
//  Laidian
//
//  Created by Tiny on 2018/2/1.
//  Copyright © tiny. All rights reserved.
//

import Foundation
import UIKit

extension UIViewController {
    
    private static var _cache : NSCache = NSCache<NSString,NSString>()
    
    private static var _storyboardIdentifierMap : [String:Set<String>] = {
        
        var map : [String:Set<String>] = [:]
        let error = safeCode {
            if let resourcePath = Bundle.main.resourcePath {
                let list = Bundle.paths(forResourcesOfType: "storyboardc", inDirectory: resourcePath)
                for path in list {
                    let name = ((path as NSString).lastPathComponent as NSString).deletingPathExtension
                    let realName = (name as NSString).components(separatedBy: "~").first!
                    let bundle = Bundle.init(path: path)
                    let nibNames : Any? = bundle?.object(forInfoDictionaryKey: "UIViewControllerIdentifiersToNibNames")
                    if let nibNamesKeys = (nibNames as? [String:Any])?.keys {
                        map[realName] = Set.init(nibNamesKeys)
                    }
                }
            }
        }
        if let e = error {
            print(e)
        }
        return map
    }()
    
    final class func instance() -> UIViewController? {
        return instance(NSStringFromClass(self))
    }
    
    final class func instance(_ identifier : String) -> UIViewController? {
        
        if let cachedName = _cache.object(forKey: identifier as NSString) {
            if let vc = tryTakeOutInstance(storyboard: cachedName as String, identifier: identifier) {
                return vc
            }
            _cache.removeObject(forKey: identifier as NSString)
        }
        
        let id = (identifier as NSString).components(separatedBy: ".").last!
        let map = _storyboardIdentifierMap
        for (name,list) in map {
            
            if list.contains(id) {
                if let vc = tryTakeOutInstance(storyboard: name, identifier: id) {
                    _cache.setObject(name as NSString, forKey: identifier as NSString)
                    return vc
                }
            }
        }
        return self.init()
    }
    
     class func tryTakeOutInstance(storyboard named : String, identifier : String) -> UIViewController? {
        var vc : UIViewController?
        let selStr = "identifier" + "To" + "Nib" + "Name" + "Map"
        let sel = NSSelectorFromString(selStr)
        let error = safeCode {
            let sb = UIStoryboard.init(name: named, bundle: Bundle.main)
            let obj = sb.perform(sel).takeUnretainedValue() as? Dictionary<String,String>
            if obj != nil && obj!.values.contains(identifier) {
                vc = sb.instantiateViewController(withIdentifier: identifier)
            }
        }
        if let e = error {
            print(e)
        }
        return vc;
    }
}
